README.md
<p align="left">
  <img src="doc/logo.png" width="30%">
</p>

# Infinity API Tools

`infinity_tools` contains modules for interacting with the [Infinity API](https://medium.com/infinity-ai/infinity-api-82149d21c87c) by [Infinity AI, Inc](https://toinfinity.ai/).

The Infinity API provides synthetic data to help ML teams build better, debug faster, and prototype at the click of a button.

# Overview

Multiple synthetic data generators are available through the Infinity API. Each generator has a distinct set of input parameters available through a RESTful API. `infinity_tools` contains wrapper libraries to simplify use of each generator and make common workflows ergonomic. There are also data science, visualization, and machine learning submodules available to further support workflows canonical for a given generator.

Currently, `infinity_tools` supports the following generator APIs:

- [VisionFit](infinity_tools/visionfit/): an exercise/fitness generator targeting computer vision applications.
- [SenseFit](infinity_tools/sensefit/): an exercise/fitness generator targeting IMU and wearable time-series applications.
- [Spills](infinity_tools/spills/): a computer vision generator that produces photorealistic videos of spills in common retail, industrial, and workplace environments.

The [common](infinity_tools/common/) submodule contains data structures and functionality common across generator-specific support libraries.

# Installation

You can use Conda to install this package.

Ensure you are using conda with the [conda-forge](https://conda-forge.org/) channel. You can check that this is the case with `conda config --show channels`. We recommend using `miniforge` for your conda installation (instructions [here](https://github.com/conda-forge/miniforge)), which sets `conda-forge` as the default channel.

## OS X (Intel) or Linux
- Create a fresh Conda environment 
    - `conda create --name infinity python=3.9`
    - `conda activate infinity`
- Make sure you have the latest pip
    - `python -m pip install --upgrade pip`
- Clone the repo 
    - `git clone git@github.com:toinfinityai/infinity-tools.git`
- Install the package into your environment
    - `pip install -e PATH_TO_INFINITY_TOOLS`

## OS X (ARM-based such as M1)
- Create a fresh Conda environment
    - `conda create --name infinity python=3.9`
    - `conda activate infinity`
- Make sure you have the latest pip
    - `python -m pip install --upgrade pip`
- Clone the repo 
    - `git clone git@github.com:toinfinityai/infinity-tools.git`
- Edit `pyproject.toml` in the `infinity-tools` repo, changing `tensorflow` to 
  `tensorflow-macos`
- Install the package into your environment
    - `pip install -e PATH_TO_INFINITY_TOOLS`
- If you received an error about `h5py`, please install it with conda
    - `conda install h5py`
- Repeat installing the package
    - `pip install -e PATH_TO_INFINITY_TOOLS`

# Swagger UI

[<img alt="Swagger UI" height="50px" src="https://static1.smartbear.co/swagger/media/assets/images/swagger_logo.svg" />](https://api.toinfinity.ai/api/schema/swagger-ui/)

A Swagger UI is available for the Infinity API [here](https://api.toinfinity.ai/api/schema/swagger-ui/). The Swagger UI provides a web-based interface to perform low-level interactions with the REST endpoint through the browser. A valid authentication token is required.

# Infinity API Tutorial Notebooks

The [Infinity API Tutorials](https://github.com/toinfinityai/infinity-tutorials) repo provides extensive notebook examples that make use of the `infinity_tools` libary. It also provides detailed documentation on API inputs and outputs, including the labels provided with synthetic data.

# Contact
The Infinity API is brought to you by [Infinity AI](https://toinfinity.ai/). We're a small team of dedicated engineers who specialize in generating custom synthetic datasets and built these API hoping they would be useful to people like you! Drop us a line at [info@toinfinity.ai](mailto:info@toinfinity.ai).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented here.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## [0.7.0] - 2022-10-11

### Added
- Added option to overlay category or action label on bounding box videos.

### Changed
- Miscellaneous updates to `visionfit.vis` to make code more generic / applicable to other generators.

### Fixed
- Use correct order of height and width with `cv2.VideoWriter` (avoids broken output videos for non-square aspect ratios).

## [0.6.0] - 2022-09-15

### Added
- Added support for querying usage statistics:
  - `infinity_tools.common.api.query_usage_datetime_range`
  - `infinity_tools.common.api.query_usage_last_n_days`

## [0.5.0] - 2022-08-16

### Added
- Added `spills` submodule to support interactions with the Spills API and related visualizations.

## [0.4.0] - 2022-08-05

### Changed

- Remove hardcoded parameter definitions/constraints and instead retrieve dynamically from the API.
- Update visualization tooling for job parameter distributions.

## [0.3.1] - 2022-07-04

### Changed

- VisionFit bumped to `v0.3.1`.

### Fixed

- Use the correct image resolution in exported camera matrices for VisionFit.

## [0.3.0] - 2022-06-13

### Added
- Support for up to 20 repetitions in both SenseFit and VisionFit.
- Added many more exercises to SenseFit and VisionFit generators.
- New VisionFit parameters:
  - `rel_baseline_speed`
  - `camera_distance`
  - `avatar_identity`
  - `relative_height`
  - `relative_weight`
- New SenseFit parameters:
  - `rel_baseline_speed`
  - `wrist_offset_deg`

### Changed
- Polling mechanism used by `await_jobs` and `get_completed_jobs` now takes advantage of new backend updates to be more efficient and use only a single HTTP request, regardless of batch size.
- Fixed/updated documentation for various parameters and their constraints.
- VisionFit bumped to `v0.3.0`.
- SenseFit bumped to `v0.2.0`.
- Maximum allowable value for `kinematic_noise_factor` increased to 2.0 for SenseFit.
- Renamed existing exercises to follow a new naming convention.
- Increased the default kinematic variation for select exercises: `PUSHUP`, `BURPEE`, and `PUSHUP-EXPLOSIVE`.

### Removed
- `seconds_per_rep` parameter removed from both VisionFit and SenseFit.

## [0.2.0] - 2022-05-25

### Added

- Added Github Action for Python `black` code formatting.
- Added 30 frames-per-second option to `frame_rate` parameter for VisionFit.
- New vertex keypoints added to the output of VisionFit.

### Changed

- Whitespace changes from applying `black` formatter.
- VisionFit bumped to `v0.2.0`.
- SenseFit bumped to `v0.1.1`.
- Serial request delay changed to parameter from hardcoded value in `await_jobs`.
- Removed `assert`s and replaced with raised `Exception`s throughout codebase.

### Fixed

- Removed broken "DUMBBELL_DEADLIFT" exercise for SenseFit.


infinity_tools/visionfit/vis.py
import glob
import json
import os
import pandas as pd
import cv2
import matplotlib.pyplot as plt
import numpy as np
import numpy.typing as npt
import plotly.graph_objects as go
import plotly.colors as colors
from pathlib import Path
from typing import Any, List, Tuple, Dict
from zipfile import ZipFile
from IPython.display import HTML, display, clear_output
from matplotlib import animation
from matplotlib.animation import FuncAnimation
from plotly.subplots import make_subplots
from pycocotools.coco import COCO
from infinity_tools.common.vis.videos import (
    parse_video_frames,
    stack_videos,
    create_grid_of_videos,
)
from infinity_tools.visionfit.datagen import apply_movenet_to_video, preprocess_frame


def unzip(zipped_folder: str, output_dir: str):
    """Unzips contents of zip file to output directory."""
    with ZipFile(zipped_folder, "r") as zip_ref:
        zip_ref.extractall(output_dir)


def visualize_all_labels(video_job_folder: str) -> str:
    """Visualizes labels for a video job returned from the API.

    Args:
        video_job_folder: Path to the folder containing the video job results.

    Returns:
        Path to resulting animation video.
    """
    output_directory = os.path.join(video_job_folder, "output")
    os.makedirs(output_directory, exist_ok=True)

    video_rgb_path = os.path.join(video_job_folder, "video.rgb.mp4")
    video_json_path = os.path.join(video_job_folder, "video.rgb.json")
    zipped_video_path = os.path.join(video_job_folder, "video.rgb.zip")
    job_json_path = os.path.join(video_job_folder, "job.json")
    job_params = json.load(open(job_json_path))["params"]

    video_rgb_extracted = os.path.join(video_job_folder, "video.rgb")
    os.makedirs(video_rgb_extracted, exist_ok=True)
    unzip(zipped_video_path, video_rgb_extracted)

    imgs = parse_video_frames(video_rgb_path)
    fps = int(job_params["frame_rate"])
    rep_count = parse_rep_count_from_json(video_json_path)
    coco = COCO(video_json_path)

    bounding_box_path = create_bounding_boxes_video(
        output_path=os.path.join(output_directory, "bounding_box.mp4"), imgs=imgs, fps=fps, coco=coco
    )
    skeleton_path = create_keypoint_connections_video(
        output_path=os.path.join(output_directory, "skeleton.mp4"), imgs=imgs, fps=fps, coco=coco
    )
    cuboids_path = create_cuboids_video(
        output_path=os.path.join(output_directory, "cuboids.mp4"), imgs=imgs, fps=fps, coco=coco
    )
    _3D_path = create_3D_keypoints_video(
        output_path=os.path.join(output_directory, "3D_keypoints.mp4"),
        fps=fps,
        coco=coco,
        width_in_pixels=imgs.shape[2],
        height_in_pixels=imgs.shape[1],
    )
    segmentation_path = create_segmentation_video(
        output_path=os.path.join(output_directory, "segmentation.mp4"),
        folder=video_rgb_extracted,
        fps=fps,
        image_width=imgs.shape[2],
        image_height=imgs.shape[1],
    )

    clear_output()

    row1 = [video_rgb_path, segmentation_path, bounding_box_path]
    label_videos_row_1 = stack_videos(row1, axis=2)
    row2 = [cuboids_path, skeleton_path, _3D_path]
    label_videos_row_2 = stack_videos(row2, axis=2)
    label_grid_path = stack_videos([label_videos_row_1, label_videos_row_2], axis=1)
    ts_path = animate_time_series(
        output_path=os.path.join(output_directory, "timeseries.mp4"),
        y_axis=rep_count,
        fps=fps,
        width_in_pixels=imgs.shape[1] * 2,
        height_in_pixels=imgs.shape[1] * 2,
    )
    return stack_videos([label_grid_path, ts_path], axis=2)


def parse_rep_count_from_json(json_path: str, rep_count_col: str = "rep_count_from_start") -> List[float]:
    return [x[rep_count_col] for x in json.load(open(json_path))["images"]]


def animate_time_series(
    output_path: str,
    y_axis: List[float],
    fps: int,
    dpi: int = 150,
    width_in_pixels: int = 300,
    height_in_pixels: int = 300,
    title: str = "Rep Count",
    xlabel: str = "Time [s]",
) -> str:
    """Generates time series animation.

    Args:
        output_path: Filename to save output video.
        y_axis: The time series to animate.
        fps: Frame rate of output video.
        width_in_pixels: Width of output video.
        height_in_pixels: Height of output video.
        dpi: DPI of output video.
        title: Title of output video.
        xlabel: X-axis label of output video.

    Returns:
        Path to resulting animation video.
    """

    num_frames = len(y_axis)
    interval = 1 / fps * 1000
    time = np.arange(num_frames) / fps

    fig_height = height_in_pixels / dpi
    fig_width = width_in_pixels / dpi
    fig, ax = plt.subplots(1, 1, figsize=(fig_width, fig_height), dpi=dpi)
    ax.plot(time, y_axis)
    ax.set_title(title)
    ax.set_xlabel(xlabel)
    ax.grid("on")

    ylim = ax.get_ylim()
    (line,) = ax.plot([0, 0], ylim, color="red")
    fig.tight_layout()

    def update(frame: int, time: npt.NDArray, ylim: Tuple[float, float]):
        line.set_data([time[frame], time[frame]], ylim)

    anim = FuncAnimation(
        fig=fig,
        func=update,
        frames=num_frames,
        fargs=(
            time,
            ylim,
        ),
        interval=interval,
        blit=False,
    )

    anim.save(output_path)
    plt.close()
    return output_path


def create_bounding_boxes_video(
    output_path: str, imgs: npt.NDArray, fps: int, coco: Any, add_text: bool = False
) -> str:
    """Overlays bounding box annotations onto video.

    Args:
        output_path: Path to output video.
        imgs: Frames of the video.
        fps: Frame rate of input video.
        coco: COCO data.
        add_text: If True, adds category or action label above each bounding box.

    Returns:
        Path to resulting animation video.
    """
    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    image_dims = (imgs.shape[2], imgs.shape[1])
    out = cv2.VideoWriter(output_path, fourcc, fps, image_dims)

    for img, img_data in zip(imgs, coco.imgs.values()):
        canvas = img.copy()
        img_id = img_data["id"]
        ann_ids = coco.getAnnIds(imgIds=img_id)
        anns = coco.loadAnns(ann_ids)
        for ann in anns:
            if "bbox" not in ann:
                continue
            x, y, w, h = tuple(np.array(ann["bbox"]).astype(int))
            cv2.rectangle(canvas, (x, y), (x + w, y + h), color=(255, 255, 255), thickness=2)
            if add_text:
                category_name = coco.cats[ann["category_id"]]["name"]
                text = ann.get("action", category_name)
                cv2.putText(
                    img=canvas,
                    text=text,
                    org=(x, y),
                    fontFace=cv2.FONT_HERSHEY_SIMPLEX,
                    fontScale=0.5,
                    color=(255, 255, 255),
                    thickness=1,
                    lineType=cv2.LINE_AA,
                )

        out.write(canvas)
    out.release()
    return output_path


def get_project_root() -> str:
    """Utility method to get the project root directory."""
    return Path(__file__).parent.parent


def get_armature_connections() -> Dict:
    """Returns armature connections from local json file."""
    project_root = get_project_root()
    armature_json = os.path.join(project_root, "common", "vis", "assets", "armature_connections.json")
    return json.load(open(armature_json, "r"))


def create_keypoint_connections_video(output_path: str, imgs: npt.NDArray, fps: int, coco: Any) -> str:
    """Overlays keypoint connection annotations onto video.

    Args:
        output_path: Path to output video.
        imgs: Frames of the video.
        coco: COCO data.
        job_params: Job parameters from the API.

    Returns:
        Path to resulting animation video.
    """
    kp_connections = get_armature_connections()
    image_dims = (imgs.shape[2], imgs.shape[1])
    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    out = cv2.VideoWriter(output_path, fourcc, fps, image_dims)

    for img, img_data in zip(imgs, coco.imgs.values()):
        canvas = img.copy()
        img_id = img_data["id"]
        ann_ids = coco.getAnnIds(imgIds=img_id)
        anns = coco.loadAnns(ann_ids)
        for ann in anns:
            if "armature_keypoints" not in ann:
                continue
            keypoints = ann["armature_keypoints"]
            for parent, child in kp_connections:
                if keypoints[parent]["z"] < 0 or keypoints[child]["z"] < 0:
                    continue
                x0 = keypoints[parent]["x"]
                y0 = keypoints[parent]["y"]
                x1 = keypoints[child]["x"]
                y1 = keypoints[child]["y"]
                cv2.line(canvas, (x0, y0), (x1, y1), color=(255, 255, 255), thickness=2)
                cv2.circle(canvas, (x0, y0), radius=4, color=(255, 255, 255), thickness=-1)

        out.write(canvas)
    out.release()
    return output_path


def create_cuboids_video(output_path: str, imgs: npt.NDArray, fps: int, coco: Any) -> str:
    """Overlays cuboid annotations onto video.

    Args:
        output_path: Path to output video
        imgs: Frames of the video
        coco: COCO data
        job_params: Job parameters from the API

    Returns:
        Path to resulting animation video.
    """
    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    image_dims = (imgs.shape[2], imgs.shape[1])
    out = cv2.VideoWriter(output_path, fourcc, fps, image_dims)
    cuboid_edges = [
        [0, 1],
        [0, 4],
        [0, 3],
        [1, 2],
        [2, 3],
        [3, 7],
        [2, 6],
        [1, 5],
        [4, 5],
        [5, 6],
        [6, 7],
        [4, 7],
    ]

    for img, img_data in zip(imgs, coco.imgs.values()):
        canvas = img.copy()
        img_id = img_data["id"]
        ann_ids = coco.getAnnIds(imgIds=img_id)
        anns = coco.loadAnns(ann_ids)
        for ann in anns:
            if "cuboid_coordinates" not in ann:
                continue
            cuboid_points = ann["cuboid_coordinates"]

            if "z" in cuboid_points[0]:
                all_in_front = all(
                    [(cuboid_points[e[0]]["z"] > 0 and cuboid_points[e[1]]["z"] > 0) for e in cuboid_edges]
                )
                if not all_in_front:
                    continue

            for edge in cuboid_edges:
                start_point = [cuboid_points[edge[0]]["x"], cuboid_points[edge[0]]["y"]]
                end_point = [cuboid_points[edge[1]]["x"], cuboid_points[edge[1]]["y"]]
                color = tuple([int(255 * x) for x in coco.cats[ann["category_id"]]["color"][::-1]])
                color = (255, 255, 255)
                canvas = cv2.line(
                    canvas,
                    tuple(start_point),
                    tuple(end_point),
                    color=color,
                    thickness=2,
                )
        out.write(canvas)
    out.release()
    return output_path


def create_2D_keypoints_video(output_path: str, imgs: npt.NDArray, fps: int, coco: Any) -> str:
    """Overlays 2D keypoints onto video.

    Args:
        output_path: Path to output video
        imgs: Frames of the video
        coco: COCO data
        job_params: Job parameters from the API

    Returns:
        Path to resulting animation video.
    """

    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    image_dims = (imgs.shape[2], imgs.shape[1])
    out = cv2.VideoWriter(output_path, fourcc, fps, image_dims)

    for img, img_data in zip(imgs, coco.imgs.values()):
        canvas = img.copy()
        img_id = img_data["id"]
        ann_ids = coco.getAnnIds(imgIds=img_id)
        anns = coco.loadAnns(ann_ids)
        for ann in anns:
            if "armature_keypoints" not in ann:
                continue
            keypoints = ann["armature_keypoints"]
            for keypoint_name, keypoint_info in keypoints.items():
                if keypoint_name == "root":
                    continue
                if keypoint_info["z"] < 0:
                    continue
                x, y = keypoint_info["x"], keypoint_info["y"]
                cv2.circle(canvas, (x, y), radius=3, color=(255, 255, 255), thickness=-1)
        out.write(canvas)
    out.release()
    return output_path


def create_3D_keypoints_video(
    output_path: str,
    fps: int,
    coco: Any,
    dpi: int = 150,
    width_in_pixels: int = 200,
    height_in_pixels: int = 200,
) -> str:
    """Visualizes 3D keypoint annotations as video.

    Args:
        output_path: Path to output video
        fps: desired frame rate of output video
        coco: COCO data
        dpi: resolution (dots per inch)
        width_in_pixels: Width of figure, in pixels
        height_in_pixels: Height of figure, in pixels

    Returns:
        Path to resulting animation video.
    """
    fig_height = height_in_pixels / dpi
    fig_width = width_in_pixels / dpi
    figsize = (fig_width, fig_height)
    fig = plt.figure(dpi=dpi, figsize=figsize)
    ax = fig.add_subplot(projection="3d")

    global_coords = []
    for img_data in coco.imgs.values():
        img_id = img_data["id"]
        ann_ids = coco.getAnnIds(imgIds=img_id)
        anns = coco.loadAnns(ann_ids)
        for ann in anns:
            if "armature_keypoints" not in ann:
                continue
            keypoints = ann["armature_keypoints"]
            for _, keypoint_info in keypoints.items():
                global_coords.append(
                    [
                        keypoint_info["x_global"],
                        keypoint_info["y_global"],
                        keypoint_info["z_global"],
                    ]
                )
    if all(["person_idx" not in ann for ann in coco.anns.values()]):
        num_people = 1
    else:
        num_people = max([ann["person_idx"] for ann in coco.anns.values() if "person_idx" in ann]) + 1
    global_coords = np.array(global_coords).reshape(-1, num_people * len(keypoints), 3)

    def update(num):
        graph._offsets3d = (
            global_coords[num, :, 0],
            global_coords[num, :, 1],
            global_coords[num, :, 2],
        )

    graph = ax.scatter(global_coords[0, :, 0], global_coords[0, :, 1], global_coords[0, :, 2])

    ax.set_box_aspect(np.ptp(global_coords, axis=(0, 1)))
    ax.set_xticks([])
    ax.set_yticks([])
    ax.set_zticks([])
    ax.grid("off")

    ani = animation.FuncAnimation(fig, update, global_coords.shape[0], blit=False, interval=1000 / fps)
    ani.save(output_path)
    plt.close()
    return output_path


def create_segmentation_video(output_path: str, folder: str, fps: int, image_width: int, image_height: int) -> str:
    """Creates a video of frame-wise segmentation masks.

    Args:
        output_path: Path to output video.
        folder: Folder which contains the per-frame data.
        job_params: Job parameters from the API.

    Returns:
        Path to resulting animation video.
    """
    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    out = cv2.VideoWriter(output_path, fourcc, fps, (image_width, image_height))
    iseg_paths = sorted(glob.glob(os.path.join(folder, "*.iseg.png")))
    for img_path in iseg_paths:
        out.write(cv2.imread(img_path))
    out.release()
    return output_path


def summarize_batch_results_as_dataframe(batch_folder: str) -> pd.DataFrame:
    """Compiles job parameters and post-job metadata associated with a batch.

    Args:
        batch_folder: Path to batch folder containing individual job results.

    Returns:
        Dataframe containing batch job parameters and any metadata extracted
        from the resulting job annotations.
    """

    def _convert_to_float(x):
        """Handler for when the job params have been converted to strings."""
        try:
            return x.astype(float)
        except:  # noqa: E722
            return x

    rgb_jsons = glob.glob(os.path.join(batch_folder, "**/video.rgb.json"), recursive=True)

    metadata = []
    for rgb_json in rgb_jsons:
        job_id = os.path.basename(os.path.dirname(rgb_json))
        full_path = os.path.join(batch_folder, job_id)
        job_params_path = os.path.join(os.path.dirname(rgb_json), "job.json")
        job_params = json.load(open(job_params_path))["params"]
        json_data = json.load(open(rgb_json))
        num_frames = len(json_data["images"])
        anns = json_data["annotations"]
        person_cat_id = {e["name"]: e["id"] for e in json_data["categories"]}["person"]
        avg_percent_in_fov = np.mean([ann["percent_in_fov"] for ann in anns if ann["category_id"] == person_cat_id])
        avg_percent_occlusion = np.mean(
            [ann["percent_occlusion"] for ann in anns if ann["category_id"] == person_cat_id]
        )
        video_metadata = {
            "job_path": full_path,
            "job_id": job_id,
            "num_frames": num_frames,
            "avg_percent_in_fov": avg_percent_in_fov,
            "avg_percent_occlusion": avg_percent_occlusion,
        }
        metadata.append({**video_metadata, **job_params})
    df = pd.DataFrame(metadata)
    df = df.apply(_convert_to_float, axis=0)
    return df


def visualize_batch_results(batch_folder: str):
    """Generates histograms of job parameters and extracted metadata for a batch."""

    def _convert_to_float(x):
        """Handler for when the job params have been converted to strings."""
        try:
            return x.astype(float)
        except:  # noqa: E722
            return x

    df = summarize_batch_results_as_dataframe(batch_folder)
    columns_to_keep = [column for column in df.columns if column not in ["job_path", "job_id", "state"]]
    df = df[columns_to_keep]
    color = colors.qualitative.Plotly[0]
    df = df.apply(_convert_to_float, axis=0)

    num_per_row = 4
    row_num = len(df.columns) // num_per_row + 1
    fig = make_subplots(rows=row_num, cols=num_per_row, subplot_titles=df.columns)

    for i, col_name in enumerate(df.columns):
        row = i // num_per_row + 1
        col = i % num_per_row + 1
        if row * col > len(df.columns):
            break

        subfig = go.Figure(data=[go.Histogram(x=df[col_name], name=col_name, nbinsx=10, marker=dict(color=color))])
        fig.add_trace(subfig.data[0], row=row, col=col)
        fig.update_layout(
            template="plotly_white",
            title=f"Summary of Batch | {Path(batch_folder).stem}",
            height=200 * row_num,
            width=300 * num_per_row,
            bargap=0.1,
            showlegend=False,
        )

    fig.show("svg")


def display_batch_results(batch_folder: str) -> None:
    """Displays batch results as dataframe."""
    metadata = summarize_batch_results_as_dataframe(batch_folder)
    display(HTML(metadata.to_html()))


def crop_pad_video(video_path: str, image_width: int = 192, image_height: int = 192) -> str:
    """Crops and pads a video to a standard size."""
    cap = cv2.VideoCapture(video_path)
    fps = int(cap.get(cv2.CAP_PROP_FPS))
    output_path = video_path.replace(".mp4", "_cropped.mp4")
    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    out = cv2.VideoWriter(output_path, fourcc, fps, (image_width, image_height))
    video_frames = parse_video_frames(video_path)
    video_frames = [preprocess_frame(frame).numpy().squeeze().astype(np.uint8) for frame in video_frames]
    for frame in video_frames:
        out.write(frame)
    out.release()
    return output_path


def overlay_movenet_on_video(video_path: str, output_path: str):
    """Applies MoveNet to video and returns rendered video path."""
    cropped_video = crop_pad_video(video_path)  # pre-crop and pad here to get aspect ratio.
    landmarks_from_video = apply_movenet_to_video(video_path)
    return visualize_landmarks(cropped_video, landmarks_from_video, output_path)


def view_batch_folder_with_movenet_overlay(batch_folder: str, num_per_row: int = 4, limit_videos: int = 12) -> str:
    """Visualizes a batch folder with movenet overlay."""
    video_paths = glob.glob(os.path.join(batch_folder, "**/*.rgb.mp4"))[:limit_videos]
    output_paths = [path.replace(".rgb.mp4", ".rgb_movenet.mp4") for path in video_paths]
    movenet_paths = [overlay_movenet_on_video(path, output) for path, output in list(zip(video_paths, output_paths))]
    return create_grid_of_videos(movenet_paths, num_per_row=num_per_row)


def visualize_landmarks(video_path: str, landmarks: List[npt.NDArray], output_path: str) -> str:
    """Visualizes MoveNet landmarks for a single video.

    Args:
        video_path: Path to video that will be processed.
        landmarks: Numpy array of shape (num_frames x 17 x 4) containing
            predicted MoveNeet keypoints.
        output_path: Path to save rendered video.

    Returns:
        output_path
    """

    KEYPOINT_PARENT_CHILD = [
        (0, 1),
        (0, 2),
        (1, 3),
        (2, 4),
        (0, 5),
        (0, 6),
        (5, 7),
        (7, 9),
        (6, 8),
        (8, 10),
        (5, 6),
        (5, 11),
        (6, 12),
        (11, 12),
        (11, 13),
        (13, 15),
        (12, 14),
        (14, 16),
    ]

    cap = cv2.VideoCapture(video_path)
    fps = int(cap.get(cv2.CAP_PROP_FPS))
    image_width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    image_height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))

    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    out = cv2.VideoWriter(output_path, fourcc, fps, (image_width, image_height))

    frame_idx = 0

    while cap.isOpened():
        success, image = cap.read()
        if not success:
            break

        for parent, child in KEYPOINT_PARENT_CHILD:
            x0 = int(image_width * landmarks[frame_idx][parent][1])
            y0 = int(image_height * landmarks[frame_idx][parent][0])
            x1 = int(image_width * landmarks[frame_idx][child][1])
            y1 = int(image_height * landmarks[frame_idx][child][0])
            cv2.line(image, (x0, y0), (x1, y1), color=(255, 255, 255), thickness=1)
            cv2.circle(image, (x0, y0), radius=2, color=(176, 129, 30), thickness=-1)
        out.write(image)
        frame_idx += 1
    out.release()

    return output_path
